### PR TITLE
Add IM event path filter

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
Per IM encoding spec, we need to filter event path from event buffers

#### Change overview
-- Pass the particular interested event Path to EventManagement, and
fetch events based upon those path.

#### Testing
-- Update unit test to filter the interested path with multiple clusters, and calculate the number
of filtered event, and compare with the expected event number.

### CODE REVIEW

**Review by jmartinez-silabs — APPROVED** [review]

File: src/app/EventManagement.cpp [hunk 1]
```diff
@@ -618,9 +604,9 @@ CHIP_ERROR EventManagement::EventIterator(const TLVReader & aReader, size_t aDep
     if (event.mPriority == apEventLoadOutContext->mPriority)
     {
         apEventLoadOutContext->mCurrentSystemTime.mValue += event.mDeltaSystemTime.mValue;
-        VerifyOrExit(apEventLoadOutContext->mCurrentEventNumber < apEventLoadOutContext->mStartingEventNumber,
+        VerifyOrExit(!(apEventLoadOutContext->mCurrentEventNumber >= apEventLoadOutContext->mStartingEventNumber &&
```
Comment (andy31415) [inline src/app/EventManagement.cpp:607]: This long condition seems hard to follow especially given the negation in front. Since either of the conditions being false results in an exit, could we split into two?

```cpp
VerifyOrExit(apEventLoadOutContext->mCurrentEventNumber < apEventLoadOutContext->mStartingEventNumber, err = CHIP_EVENT_ID_FOUND)

VerifyOrExit(!IsInterestedEventPaths(apEventLoadOutContext->mpInterestedEventPaths, event), err = CHIP_EVENT_ID_FOUND);
```

Readability-wise I find this logic somewhat unclear: why do we say "I found the value" if the check is "Is this event interesting"?

Comment (andy31415) [reply src/app/EventManagement.cpp:607]: Sinc the exit branch has no cleanup part, consider using "return" instead of exit to make the code easier to follow and less macro-heavy. A `if (!IsInterested...) return CHIP_EVENT_ID_FOUND` seems easier to read than a macro expansion.

File: src/app/EventManagement.cpp [hunk 2]
```diff
@@ -691,6 +678,21 @@ CHIP_ERROR EventManagement::FetchEventsSince(TLVWriter & aWriter, ClusterInfo *
     return err;
 }
 
+bool EventManagement::IsInterestedEventPaths(ClusterInfo * apInterestedEventPaths, EventEnvelopeContext & aEvent)
```
Comment (andy31415) [inline src/app/EventManagement.cpp:681]: I am not sure "IsInterested" conveys what this method is doing. The logic seems to be "Loop over a chain of cluster info and determine if an event with the exact node/endpoint/cluster/eventid is contained within that chain".

Consider "ContainsPath" (assuming we call node/ep/cluster/event a "path") or event.

Can this be a static method within the cpp file instead of being a static method inside the class?
Additional nit: parameter naming feels odd. The "a" and "ap" previx makes parameters long and the names are not as matched "ClusterInto  *clusterInfoList" and "EventEnvelopeContext &context" seems to be clearer.

EventEnvelopeContext should be `const`

File: src/app/EventManagement.h [hunk 3]
```diff
@@ -188,6 +188,25 @@ enum class EventManagementStates
     Shutdown   = 3  // Not capable of performing any logging operation
 };
 
+/**
+ * @brief
+ *  Internal structure for traversing events.
+ */
+struct EventEnvelopeContext
```
Comment (andy31415) [inline src/app/EventManagement.h:195]: Do we need to move this inside the header?

File: src/app/EventManagement.cpp (hunk 1 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:607]: Yep, just move this logic into IsInterested function, and update it with return pattern instead of exit.

File: src/app/EventManagement.cpp (hunk 2 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:681]: Done, thanks

File: src/app/EventManagement.h (hunk 3 above)
Comment (yunhanw-google) [reply src/app/EventManagement.h:195]: no need now, thanks, resovled.

File: src/app/EventManagement.cpp [hunk 4]
```diff
@@ -604,27 +628,30 @@ CHIP_ERROR EventManagement::EventIterator(const TLVReader & aReader, size_t aDep
 
     innerReader.Init(aReader);
     err = innerReader.EnterContainer(tlvType);
-    SuccessOrExit(err);
+    ReturnErrorOnFailure(err);
```
Comment (andy31415) [inline src/app/EventManagement.cpp:631]: nit: can be

```cpp
ReturnErrorOnFailure(innerReader.EnterContainer(tlvType));
```

File: src/app/EventManagement.cpp [hunk 5]
```diff
@@ -604,27 +628,30 @@ CHIP_ERROR EventManagement::EventIterator(const TLVReader & aReader, size_t aDep
...
 
     if (event.mPriority == apEventLoadOutContext->mPriority)
     {
         apEventLoadOutContext->mCurrentSystemTime.mValue += event.mDeltaSystemTime.mValue;
-        VerifyOrExit(apEventLoadOutContext->mCurrentEventNumber < apEventLoadOutContext->mStartingEventNumber,
-                     err = CHIP_EVENT_ID_FOUND);
-        apEventLoadOutContext->mCurrentEventNumber++;
+        if(IsInterestedEventPaths(apEventLoadOutContext, event))
+        {
+            return CHIP_EVENT_ID_FOUND;
```
Comment (andy31415) [inline src/app/EventManagement.cpp:650]: This makes me wonder why we return something that is not 'CHIP_ERROR_' prefixed as a CHIP_ERROR.

it looks like we are trying to have two return values: real error and some status of found or not. This seems slightly counter-intuitive and maybe we could design our API differently.

Saying "Error of event id was found" sounds wrong, so it is not a CHIP_ERROR.

**Review by andy31415 — APPROVED** [review]

File: src/app/EventManagement.cpp (hunk 5 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:650]: https://github.com/project-chip/connectedhomeip/issues/7141
file this issue for the investigation, thanks

File: src/app/EventManagement.cpp (hunk 4 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp]: done, thanks

**Review by msandstedt — APPROVED** [review]

File: src/app/EventManagement.cpp [hunk 6]
```diff
@@ -603,28 +627,30 @@ CHIP_ERROR EventManagement::EventIterator(const TLVReader & aReader, size_t aDep
...
-    SuccessOrExit(err);
+    ReturnErrorOnFailure(err);
 
     if (event.mPriority == apEventLoadOutContext->mPriority)
     {
         apEventLoadOutContext->mCurrentSystemTime.mValue += event.mDeltaSystemTime.mValue;
-        VerifyOrExit(apEventLoadOutContext->mCurrentEventNumber < apEventLoadOutContext->mStartingEventNumber,
-                     err = CHIP_EVENT_ID_FOUND);
-        apEventLoadOutContext->mCurrentEventNumber++;
+        if (IsInterestedEventPaths(apEventLoadOutContext, event))
```
Comment (msandstedt) [inline src/app/EventManagement.cpp:647]: It really seems like this could be better than O(n^2), but it appears it will work correctly.
